### DISCUSSION TITLE
Support for native PyTorch Mixed precision training

### DISCUSSION THREAD

**Discussion by rezafuru** [yoshitomo-matsubara/torchdistill#247] [Ideas]
Hi,

I was wondering whether there are any plans for supporting native mixed precision training? The native mixed precision training fixes quite a lot of pain points of Apex. Moreover, apex has entered [maintenace mode](https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/). 

It does not seem like torchdistill has great support for Apex in the first place, since it does not store/load the scaler for checkpoints and only works with optimization level `O1`. 

I've already adjusted the training box classes in torchdistill locally to work with native MP, and in case you're interested I could extend my implementation and create a PR. In that case, I'd like to know whether a breaking change would be okay, i.e. whether I can remove apex alltogether. The break would be minor, as it would only affect users which were currently using apex and torchdistill together.

BR
Alireza

**Comment by yoshitomo-matsubara** [comment]
Hi @rezafuru 

Thank you for the suggestion!
I feel that many users do not use the apex feature with torchdistill, and actually I was planning to exclude Apex from the package at some point since torchdistill also supports [Hugging Face's Accelerate](https://github.com/huggingface/accelerate) (See [one of my example scripts](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/hf_transformers/text_classification.py))

If you think the native PyTorch Mixed precision training would be (at least) as helpful for users as HF's Accelerate, I will be happy to review the PR

**Reply by rezafuru** [reply]
Hi,

I was not aware of the Accelerator support. I'll compare the native MP and hugging face accelerator and see whether it is still worth bothering with a PR

BR
Alireza

**Reply by rezafuru** [reply]
The Hugging Face Accelerator is amazing, not sure how I missed out on this one :D. No point in integrating pytorch's native MP training imho.  

However, I'd still argue to remove apex alltogether from torchdistill as it is on maintenance mode and seems to cause more headache than it is worth. 

BR
Alireza

**Reply by yoshitomo-matsubara** [reply]
Hi @rezafuru 

I will plan to remove apex from torchdistill. So, feel free to close this discussion
Thanks for the suggestion!
